File: cpp/429.cpp
#include <iostream>
#include <vector>
#include <queue>

using namespace std;

// Definition for a Node.
class Node {
public:
    int val;
    vector<Node*> children;

    Node() {}

    Node(int _val) {
        val = _val;
    }

    Node(int _val, vector<Node*> _children) {
        val = _val;
        children = _children;
    }
};

class Solution {
public:
    vector<vector<int>> levelOrder(Node* root) {
      vector<vector<int>> ret;
      if (!root)
          return ret;

      queue<Node *> q;
      q.push(root);
      while (!q.empty()) {
          int size = q.size();
          ret.push_back(vector<int>());
          while (size) {
              Node *cur = q.front();
              ret.back().push_back(cur->val);
              int children = cur->children.size();
              int cnt = 0;
              while (cnt < children)
                  q.push(cur->children[cnt++]);
              q.pop();
              size--;
          }
      }

      return ret;
    }
};

File: go/429.go
package main


// Definition for a Node.
type Node struct {
    Val int
    Children []*Node
}


func levelOrder(root *Node) [][]int {
    var ret [][]int = make([][]int, 0)
    if (root == nil) {
        return ret
    }
    var q []*Node = make([]*Node, 0)
    q = append(q, root)
    for len(q) > 0 {
        var size int = len(q)
        ret = append(ret, make([]int, 0))
        for size > 0 {
            var cur *Node = q[0]
            ret[len(ret)-1] = append(ret[len(ret)-1], cur.Val)
            q = q[1:]
            size--
            for _, c := range cur.Children {
                q = append(q, c)
            }
        }
    }
    return ret
}
